x/rollapp/transfergenesis/ibc_module.go

```diff
@@ -0,0 +1,258 @@
+	l := w.logger(ctx, packet)
+
+	l.Debug("Processing inbound packet.")
+
+	if !w.delayedackKeeper.IsRollappsEnabled(ctx) {
+		l.Debug("Rollapps are disabled.")
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	transfer, err := w.rollappKeeper.GetValidTransfer(ctx, packet.GetData(), packet.GetDestPort(), packet.GetDestChannel())
+	if err != nil {
+		l.Error("Get valid transfer from received packet", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "transfer genesis: get valid transfer"))
+	}
+
+	if !transfer.IsRollapp() {
+		l.Debug("Transfer is not from a rollapp.")
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	memo, err := getMemo(transfer.GetMemo())
+	if errorsmod.IsOf(err, gerr.ErrNotFound) {
+		l.Debug("Memo not found.")
+		// If someone tries to send a transfer without the memo before the bridge is open, they will
+		// be blocked at the transfersenabled middleware
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+	if err != nil {
+		l.Error("Get memo.", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "get memo"))
+	}
+
+	ra := transfer.Rollapp
+
+	nTransfersDone, err := w.rollappKeeper.VerifyAndRecordGenesisTransfer(ctx, ra.RollappId, memo.ThisTransferIx, memo.TotalNumTransfers)
+	if errorsmod.IsOf(err, derr.ErrViolatesDymensionRollappStandard) {
+		// The rollapp has deviated from the protocol!
+		handleFraudErr := w.handleFraud(ra.RollappId)
+		if err != nil {
+			l.Error("Handling fraud.", "err", handleFraudErr)
+		} else {
+			l.Info("Handled fraud: verify and record genesis transfer.", "err", err)
+		}
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "transfer genesis"))
+	}
+	if err != nil {
+		l.Error("Verify and record transfer.", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "transfer genesis: verify and record"))
+	}
+
+	// it's a valid genesis transfer!
+
+	err = w.registerDenomMetadata(ctx, ra.RollappId, ra.ChannelId, memo.Denom)
+	if err != nil {
+		l.Error("Register denom metadata.", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "transfer genesis: register denom metadata"))
+	}
+
+	l.Debug("Received valid genesis transfer. Registered denom data.",
+		"num total", memo.TotalNumTransfers,
+		"num received so far", nTransfersDone,
+		"this transfer ix", memo.ThisTransferIx,
+	)
+
+	if nTransfersDone == memo.TotalNumTransfers {
+		// The transfer window is finished! Queue up a finalization
+		ctx.EventManager().EmitEvent(allTransfersReceivedEvent(ra.RollappId, nTransfersDone))
+		w.rollappKeeper.EnableTransfers(ctx, ra.RollappId)
+		l.Info("All genesis transfers received, bridge opened.",
+			"rollapp", ra.RollappId,
+			"n transfers", nTransfersDone)
+	}
+
+	l.Debug("Passing on the transfer down the stack, but skipping delayedack, the transferEnabled blocker and bridging fee.")
+
+	return w.IBCModule.OnRecvPacket(
```

Comment:
not sure why u need three context wrapping with same value?
don't u just need to set `ctx.WithValue` here?

Reply:
They're different


type (
	key1 struct{}
	key2 struct{}
)

func TestFoo(t *testing.T) {
	ctx := sdk.NewContext(nil, tmproto.Header{}, false, nil)
	ctx = ctx.WithValue(key1{}, true)
	_, ok := ctx.Value(key1{}).(bool)
	require.True(t, ok)
	_, ok = ctx.Value(key2{}).(bool)
	require.False(t, ok)
}

Reply:
why not using common struct for a key? instead of defining key in each packet?

Reply:
not sure I got you
this is the idiomatic way to add keys to the sdk context
each ibc module has it's own skip key
if there was only one everything would be coupled / the dependency graph would be messed up